packages/component/src/Activity2/Timestamp.js

```diff
@@ -0,0 +1,83 @@
+import { connect } from 'react-redux';
+import React from 'react';
+
+import String, { getString } from '../Localization/String';
+import Context from '../Context';
+import TimeAgo from '../Utils/TimeAgo';
+
+// TODO: We could refactor this into a general component
+function sendFailed(language, replace) {
+  const text = getString('Send failed, {retry}', language);
+  const retry = getString('retry', language);
+  const match = /\{retry\}/.exec(text);
+
+  if (match) {
+    return (
+      <React.Fragment>
+        { text.substr(0, match.index) }
+        { replace(retry) }
+        { text.substr(match.index + match[0].length) }
+      </React.Fragment>
+    );
+  } else {
+    return text;
+  }
+}
+
+class Timestamp extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.handleRetryClick = this.handleRetryClick.bind(this);
+  }
+
+  handleRetryClick(evt) {
+    evt.preventDefault();
+
+    this.props.postActivity(this.props.activity);
+
+    // After clicking on "retry", the button will be gone and focus will be lost (back to document.body)
+    // We want to make sure the user stay inside Web Chat
+    this.props.focusSendBox();
+  }
+
+  render() {
+    const {
+      activity: { channelData: { state } = {}, timestamp },
+      language,
+      styleSet
+    } = this.props;
+
+    return (
+      <span className={ styleSet.timestamp }>
+        {
+          state === 'sending' ?
```

Comment:
Lots of these strings throughout the code - should this be a consts list somewhere? Same with the message types.